custom/icds_reports/sqldata/exports/poshan_progress_report.py

```diff
@@ -0,0 +1,218 @@
+    "% of trimester three women counselled on immediate and EBF",
+    "Height Measurement Efficiency",
+    "% of children between 6 months -3 years, P&LW provided THR for atleast 21+ days",
+    "% of children between 3-6 years provided SNP for atleast 21+ days"]
+
+COLS_SUMMARY = [
+    'state_name', 'district_name', 'num_launched_districts', 'num_launched_blocks', 'num_launched_awcs',
+    'avg_days_awc_open_percent', 'visits_percent', 'pse_attended_21_days_percent', 'weighed_percent',
+    'counsel_immediate_bf_percent', 'height_measured_in_month_percent', 'thr_percent',
+    'lunch_count_21_days_percent'
+]
+
+
+class PoshanProgressReport(object):
+    title = 'Poshan Progress Report'
+
+    def __init__(self, config, loc_level=0, beta=False, show_test=False):
+        self.config = config
+        self.loc_level = loc_level
+        self.beta = beta
+        self.show_test = show_test
+        self.layout = self.config['report_layout']
+        self.report_type = self.config['data_period']
+        self.row_constants = [HEADERS_COMPREHENSIVE[:], COLS_COMPREHENSIVE[:], HEADERS_SUMMARY[:], COLS_SUMMARY[:]]
+        if self.report_type == 'quarter':
+            self.quarter = self.config['quarter']
+            self.year = self.config['year']
+        if loc_level == 1:
+            self.row_constants[0].remove('District Name')
+            self.row_constants[1].remove('district_name')
+            self.row_constants[2].remove('District Name')
+            self.row_constants[3].remove('district_name')
+
+    def _generate_quarter_years(self):
+        months = []
+        end_month = self.quarter * 3
+        for i in range(end_month - 2, end_month + 1):
+            months.append(date(self.year, i, 1))
+        return months
+
+    def quarter_wise(self, filters, order_by, aggregation_level):
+        """
+        :param filters: quaterwise filter (months in [month array])
+        :param order_by: order by columns
+        :return: excel_rows
+        """
+        headers_comprehensive, cols_comprehensive, headers_summary, cols_summary = self.row_constants
+        query_set = PoshanProgressReportView.objects.filter(**filters).order_by(*order_by)
+
+        # it used to uniquely identify the row
+        # for district wise data we need district id
+        # for statewise we need state id
+        unique_id = 'district_id'
+        if aggregation_level == 1:
+            unique_id = 'state_id'
+        cols = cols_comprehensive[:]
+        cols.append(unique_id)  # used as key for the dict
+        data = query_set.values(*cols)
+        row_data_dict = {}
+        dummy_row = ['' for _ in range(0, aggregation_level)] + [0 for _ in range(0, len(
+            headers_comprehensive) - aggregation_level)]
+        headers = headers_comprehensive
+        # prepare a dict with { 'unique_id' : dummy_row}
+        del filters['month__in']
+        unique_ids = AwcLocation.objects.filter(**filters).order_by(*order_by).values(unique_id)
```

Comment:
it does not make sense to query the database again to pull state_id/district_id where you already have it on your previous query, you could use that query here

Reply:
updated